torch/csrc/distributed/rpc/rpc_with_autograd.cpp

```diff
@@ -0,0 +1,161 @@
+#include <torch/csrc/distributed/rpc/rpc_with_autograd.h>
+#include <torch/csrc/byte_order.h>
+#include <torch/csrc/distributed/rpc/utils.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+constexpr int kAutogradMessageSize = 17;
+
+AutogradMetadata::AutogradMetadata(
+    int64_t autogradContextId_,
+    int64_t autogradMessageId_)
+    : autogradContextId(autogradContextId_),
+      autogradMessageId(autogradMessageId_) {}
+
+RpcWithAutograd::RpcWithAutograd(
+    MessageType messageType,
+    const AutogradMetadata& autogradMetadata,
+    std::unique_ptr<RpcCommandBase> wrappedRpc)
+    : messageType_(messageType),
+      autogradMetadata_(autogradMetadata),
+      wrappedRpc_(std::move(wrappedRpc)) {
+  TORCH_INTERNAL_ASSERT(wrappedRpc_ != nullptr, "wrappedRpc cannot be null!");
+  TORCH_INTERNAL_ASSERT(
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_REQ ||
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_RESP);
+  wrappedMessage_ = wrappedRpc_->toMessage();
+  tensors_ = wrappedMessage_.tensors();
+  wrappedMessageType_ = wrappedMessage_.type();
+}
+
+RpcWithAutograd::RpcWithAutograd(
+    MessageType messageType,
+    const AutogradMetadata& autogradMetadata,
+    std::unique_ptr<RpcCommandBase> wrappedRpc,
+    MessageType wrappedMessageType,
+    std::vector<torch::Tensor> tensors)
+    : messageType_(messageType),
+      autogradMetadata_(autogradMetadata),
+      wrappedRpc_(std::move(wrappedRpc)),
+      wrappedMessageType_(wrappedMessageType),
+      tensors_(std::move(tensors)) {
+  TORCH_INTERNAL_ASSERT(wrappedRpc_ != nullptr, "wrappedRpc cannot be null!");
+  TORCH_INTERNAL_ASSERT(
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_REQ ||
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_RESP);
+}
+
+Message RpcWithAutograd::toMessage() {
+  auto payload = wrappedMessage_.movePayload();
+  TORCH_INTERNAL_ASSERT(!payload.empty());
+
+  // We append the message type (1 byte), autograd context id(8 bytes) and
+  // autograd message id(8 bytes) to the original message in network byte order
```

Comment:
I think this is a good example of code becoming more complicated because autograd messages wrap their non-autograd equivalents. If autograd was just integrated into the base structures, you could just have one message type at the beginning, none of this appending.

Reply:
OK, I'm actually even more surprised than I initially was.

My understanding of the original design of the system was that the network protocol and framing would be handled by whatever underlying RPC library we were using (e.g., Thrift inside Facebook). In that situation, I expect to see us construct some sort of higher-level (but maybe unstructured data-structure), and then the exact details of the binary format on the wire. But what this code is doing here is manually poking at the binary data itself; i.e., implementing another binary serialization layer on top of a message that is already going to go through a binary serialization by the underlying mechanism. Wouldn't it be cleaner to have just added space in Message to record this information, instead of the manual grunging about with bytes here?

Reply:
The original comments on the first version of this PR was trying to avoid leaking distributed autograd related code to the RPC `send` function, and those logic can be handled in the Message and request callback implementations. The reason for that is we treat RPC `send` as a pure communication layer API and distributed autograd related things should be layered on top.

I'd prefer not adding distributed autograd related fields to `Message`, as that would force every `RpcAgent` implementation to be aware of autograd stuff, while they should care of send and receive message. However, I think it is possible to keep RPC `send` concise and and at the same time avoid wrapping a message twice as well. One option could be having dedicated message types for  `SCRIPT_CALL_WITH_GRAD`, `SCRIPT_RET_WITH_GRAD`, `PYTHON_CALL_WITH_GRAD`, and `PYTHON_RET_WITH_GRAD`, and implement their `toMessage` and `fromMessage` accordingly. Or we could add a field to existing `ScriptCall`, `ScriptRet`, `PythonCall`, and `PythonRet` (the latter two requires @zhaojuanmao's PR #26314), and modify there `toMessage`, `fromMessage` accordingly. In this way, the `Message` would stay as a byte array + a tensor table.

Reply:
commented in the thread above